Add AArch64 sha1 and sha2 extensions support

linkmauve opened the pull request:
Enables https://github.com/RustCrypto/asm-hashes/pull/10

`getauxval()` is extremely cheap, as it is just a pointer to the beginning of the stack, so it doesn’t make much sense to use a `lazy_static!()` to remember the flags (at least I can’t see any difference while profiling).

newpavlov committed:
First commit

newpavlov committed:
Left only compression functions and removed everything else

newpavlov committed:
Relicensed under MIT only

newpavlov committed:
Bump whirlpool-asm to 0.2.2

newpavlov committed:
Bumped generic-array to v0.8

newpavlov committed:
Removed generic-array dep fro md5 and sha1

rubdos committed:
gcc 0.3 -> cc 1.0

newpavlov committed:
Merge pull request #2 from rubdos/cc

gcc 0.3 -> cc 1.0

newpavlov committed:
bumped versions

newpavlov committed:
added readme

newpavlov committed:
removed generic-array dependency

elliottslaughter committed:
Fix for link error on macOS. (#4)

newpavlov committed:
md5 v0.4.2, sha1 v0.4.2, sha2 v0.5.1, whirlpool v0.5.1

newpavlov committed:
cfg_checks (#6)

froydnj committed:
fix compilation issues with `cargo bench` (#7)

linkmauve committed:
Add AArch64 sha1 extension support

Here are some benchmarks on my Nintendo Switch:

Without this patch (or without `--features=asm`):
```
test bench1_10    ... bench:          66 ns/iter (+/- 0) = 151 MB/s
test bench2_100   ... bench:         497 ns/iter (+/- 1) = 201 MB/s
test bench3_1000  ... bench:       4,799 ns/iter (+/- 16) = 208 MB/s
test bench4_10000 ... bench:      47,646 ns/iter (+/- 134) = 209 MB/s
```

With this patch:
```
test bench1_10    ... bench:          32 ns/iter (+/- 1) = 312 MB/s
test bench2_100   ... bench:         165 ns/iter (+/- 1) = 606 MB/s
test bench3_1000  ... bench:       1,480 ns/iter (+/- 5) = 675 MB/s
test bench4_10000 ... bench:      14,246 ns/iter (+/- 26) = 701 MB/s
```

There are probably scheduling fixes to be made, I haven’t paid any attention to latency.

linkmauve committed:
Add AArch64 sha2 extension support

Here are some benchmarks on my Nintendo Switch:

Without this patch (or without `--features=asm`):
```
test bench1_10    ... bench:          98 ns/iter (+/- 0) = 102 MB/s
test bench2_100   ... bench:         815 ns/iter (+/- 4) = 122 MB/s
test bench3_1000  ... bench:       7,980 ns/iter (+/- 76) = 125 MB/s
test bench4_10000 ... bench:      79,373 ns/iter (+/- 793) = 125 MB/s
```

With this patch:
```
test bench1_10    ... bench:          30 ns/iter (+/- 0) = 333 MB/s
test bench2_100   ... bench:         167 ns/iter (+/- 0) = 598 MB/s
test bench3_1000  ... bench:       1,554 ns/iter (+/- 7) = 643 MB/s
test bench4_10000 ... bench:      15,231 ns/iter (+/- 88) = 656 MB/s
```

Only the SHA-256 function is implemented, my console doesn’t support the
sha512 extension.

linkmauve committed:
Try to enable Travis on ARM

tarcieri committed:
Merge pull request #10 from linkmauve/aarch64

Add AArch64 sha1 and sha2 extension support

linkmauve committed:
Disable SHA-512 benchmark on AArch64

It isn’t implemented yet.

tarcieri commented:
> it doesn’t make much sense to use a lazy_static!()

Agreed, IMO `lazy_static` is effectively obsolete at this point

linkmauve committed:
Avoid clobbering v8-v15 registers

In the AArch64 ABI, v8-v15 are callee-saved, but only their lower
64 bits, which means we were inadvertantly overwriting part of their
values.

This patch entirely removes the usage of these registers, we have plenty
of them to work with!
